polygeist/tools/cgeist/driver.cc

```diff
@@ -940,19 +978,29 @@ int main(int argc, char **argv) {
   mlir::MLIRContext context;
   registerDialects(context, syclKernelsOnly);
 
-  // Generate MLIR for the input files (in a single module).
-  mlir::OwningOpRef<mlir::ModuleOp> module(
-      mlir::ModuleOp::create(mlir::OpBuilder(&context).getUnknownLoc()));
+  // Generate MLIR for the input files.
+  auto modules = buildModules(context);
+  auto &module = modules.first;
+  auto &deviceModule = modules.second;
   llvm::DataLayout DL("");
   llvm::Triple triple;
-  processInputFiles(inputFileNames, inputCommandArgs, context, module, DL, triple, argv[0],
-                    syclKernelsOnly);
+  processInputFiles(inputFileNames, inputCommandArgs, context, module,
+                    deviceModule, DL, triple, argv[0], syclKernelsOnly);
 
   LLVM_DEBUG({
     llvm::dbgs() << "Initial MLIR:\n";
     module->dump();
   });
 
+  // For now, we will work on the device code if SYCL kernels are found and on
+  // the host code otherwise.
+  //
   // Lower the MLIR to LLVM IR, compile the generated LLVM IR.
-  return compileModule(module, context, DL, triple, LinkageArgs, argv[0]);
+  return hasKernels(*deviceModule)
+             ? compileModule(deviceModule, context, DL, triple, LinkageArgs,
+                             argv[0],
+                             {func::FuncOp::getOperationName(),
+                              gpu::GPUFuncOp::getOperationName()})
+             : compileModule(module, context, DL, triple, LinkageArgs, argv[0],
+                             {func::FuncOp::getOperationName()});
```

Comment:
Suggest to call the pipeline on both modules like this: 
```
  compileModule<mlir::ModuleOp>(module, context, DL, triple, LinkageArgs, ...);
  compileModule<mlir::GPUModuleOp>(deviceModule, context, DL, triple, LinkageArgs, ...);    
```
What do you think?

Reply:
Thanks for all your comments. I've also addressed this one.